opentelemetry-proto/tests/grpc_build.rs

```diff
@@ -41,9 +60,27 @@ fn build_grpc() {
     );
 }
 
-fn build_content_map() -> HashMap<PathBuf, String> {
-    std::fs::read_dir("src/proto/grpcio")
-        .expect("cannot open dict of generated grpc files")
+#[test]
+fn build_tonic() {
+    let before_build = build_content_map(TONIC_PROTO_DIR);
+
+    tonic_build::configure()
+        .build_server(true)
+        .build_client(true)
+        .out_dir(TONIC_PROTO_DIR)
+        .compile(TONIC_PROTO_FILES, TONIC_INCLUDES)
+        .expect("cannot compile protobuf using tonic");
+    let after_build = build_content_map(TONIC_PROTO_DIR);
+    // see build_grpc for why we cannot use assert! here.
+    assert!(
+        before_build == after_build,
+        "generated file has changed, please commit the change file and rerun the test"
+    );
+}
```

Comment:
This is slightly suboptimal: if I am reading this correctly, the code will `touch` generated files even if they are unchanged. And the problem with `touch` would be that it'd cause `Cargo` to re-build stuff (cargo looks at mtime, not at the file hash). Ideally, you'd want a pattern like this her: 

https://github.com/rust-lang/rust-analyzer/blob/d9e462fd97fe56a63da62d11bb5c4d2132e4c74a/crates/sourcegen/src/lib.rs#L171-L194

Reply:
Oh, forgot the second problem: in some CI setups, the directory with sources is read-only. The lazy update a-la `ensure_file_contents` works with that nicely -- only failing tests want to mess with src.

Reply:
Ah good point. I revise the PR to generate the files in temp dir first and then move to the src if needed